### PR TITLE
Document 1M-token training and add a context-parallelism example

### PR DESCRIPTION
**Pull request by qgallouedec**
Training on sequences far longer than the usual few thousand tokens:

- `docs/source/long_context_training.md`, a new how-to guide, moved out of `distributing_training.md` and extended with the configurations I verified and the levers in the order you hit them
- `examples/sft_qwen3_8b_1m_context/`, a runnable example that trains a book-length sequence per step on one 8xH100 node

### Measured

One 8xH100 node, bf16, `per_device_train_batch_size=1`, `loss_type="chunked_nll"` (the default). Qwen3-8B at 1,048,576 tokens: **380 s/step, 56.2 GB per GPU**. Runs end to end, 12 steps, and saves.

### Depends on

transformers >= 5.16 for gradient checkpointing `offload` (huggingface/transformers#48444, merged). Without it Qwen3-8B at 1M does not fit. The example checks the version at startup.

### Corrections since the first version

The step times I first posted came from machines where the system CUDA shadowed the cuDNN bundled with torch, so torch fell back to FlashAttention and every number was inflated. huggingface/transformers#48163 looked like a 1.7x win there and is now closed: PyTorch already prefers cuDNN on Hopper by itself.

Earlier revisions required huggingface/accelerate#4175 instead. Measured back to back on one node, the two are the same (379.75 s vs 379.76 s), so the example uses the transformers one and needs no unreleased accelerate.

---

> [!NOTE]
> **Low Risk**
> Documentation and an optional training example only; no changes to TRL training library code paths.
> 
> **Overview**
> This PR **splits long-context training out of the distributing-training guide** into a dedicated how-to, **Training Beyond 1M Tokens** (`long_context_training.md`), and wires it into the docs TOC. **`distributing_training.md`** now only points readers to that page instead of hosting the full Ring Attention / Ulysses material.
> 
> The new guide **keeps the moved CP/SP content** and **adds a “Training at 1M tokens and beyond” section**: verified step times and memory for Qwen3 at 1M–4M tokens, multi-node scaling notes, and ordered levers (`chunked_nll`, gradient-checkpointing **offload**, YaRN RoPE in `rope_parameters`), plus constraints (full attention only, no packing).
> 
> It also adds **`examples/sft_qwen3_8b_1m_context/`** — an 8-GPU FSDP2 context-parallel Accelerate config and an `SFTTrainer` script for **1,048,576-token** Qwen3-8B on PG-19 text, with a runtime check for transformers’ checkpointing `offload`. The example is listed in **`example_overview.md`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 18a98ac005f2f20c5634d360f833c204abcceaf7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: docs/source/distributing_training.md [hunk 1]
```diff
@@ -214,6 +214,70 @@ accelerate launch --config_file context_parallel_2gpu.yaml train.py
...
+   ```
+
+Context parallelism can only express full causal attention, which constrains what it can train:
+
+- **Full-attention models only.** Models with sliding-window or chunked attention layers cannot be used: their
+  per-layer mask has to be dropped, which would silently turn those layers into full causal attention.
+  Accelerate rejects such models. This rules out much of the current crop — gpt-oss (every other layer),
+  Gemma 3 and Gemma 4 (five of every six), Muse-Glimmer (39 of 52), Mistral and Ministral (every layer).
+- **No packing**, for the same reason: packed documents are kept apart by a block-diagonal mask, which is
+  dropped too. TRL raises for this combination.
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply docs/source/distributing_training.md]: The raise is #6843, which also rewrites this doc section. This PR assumes it and lands after.

File: examples/sft_long_context/sft_long_context.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+one holds 131,072 tokens and attention is computed as a ring.
+
+Example:
+
+accelerate launch \
+    --config_file examples/accelerate_configs/context_parallel_8gpu.yaml \
+    examples/sft_long_context/sft_long_context.py \
+    --model_name_or_path Qwen/Qwen3-8B \
+    --dataset_name emozilla/pg19 \
+    --max_length 1048576 \
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/sft_long_context/sft_long_context.py]: Stale: since 169a4ca0, build_dataset concatenates PG-19 books into 6M-char rows (~1.5M tokens) before truncation, so every row fills the 1M window.

File: examples/sft_long_context/sft_long_context.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+from trl import ModelConfig, ScriptArguments, SFTConfig, SFTTrainer, TrlParser
+
+
+def main(script_args, training_args, model_args):
+    # A base model used far beyond its trained context needs RoPE scaling: without it, loss at 1M starts
+    # around 10.6 instead of 4.4. In Transformers v5 `rope_theta` lives inside `rope_parameters`, so the
+    # override has to carry it.
+    training_args.model_init_kwargs = {
+        "dtype": torch.bfloat16,
+        "rope_parameters": {
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/sft_long_context/sft_long_context.py]: Real: Qwen3 on transformers 4.56/4.57 only knows rope_scaling/rope_theta, so the rope_parameters override is silently inert there. Fixed in 3132df9f by requiring transformers>=5.0 in the script deps instead of shipping dual-schema code.

File: examples/sft_long_context/sft_long_context.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    --gradient_checkpointing False \
+    --bf16 \
+    --output_dir Qwen3-8B-1M
+
+Measured on 8xH100 (step time, peak memory per GPU):
+
+    Qwen3-0.6B        137 s   27.9 GB
+    Qwen3-8B          364 s   56.2 GB
+    Qwen3-30B-A3B     483 s   46.0 GB
+    Qwen3-32B        1295 s   60.8 GB
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/sft_long_context/sft_long_context.py]: The current revision states this: the docstring marks Qwen3-32B as needing the MLP tiling (lever 5 in the guide), and the script only advertises the 8B command.

File: docs/source/distributing_training.md [hunk 5]
```diff
@@ -214,6 +214,70 @@ accelerate launch --config_file context_parallel_2gpu.yaml train.py
...
+
+   Both are speed-neutral and save several GB (grad-norm deviation < 0.5%).
+4. **For ≥ 30B models, add parameter/optimizer CPU offload** (`fsdp_offload_params: true`). Model states (~12 bytes/param sharded) move to host; at 1M-token step times the gather traffic is negligible.
+5. **For the largest dense models, tile the MLP over the sequence.** Inside a checkpointed layer, the `[seq/cp, intermediate]` gate/up projections are the last large transient, and activation offload cannot reach them — they are created and consumed within a single recompute. Splitting the MLP forward into sequence tiles caps them at `[tile, intermediate]`; the backward recomputes the same tiled code, so the bound holds there too. This is what makes Qwen3-32B fit at 1M:
+
+   ```python
+   from transformers.models.qwen3.modeling_qwen3 import Qwen3MLP
+
+   forward, tile = Qwen3MLP.forward, 16384
+   Qwen3MLP.forward = lambda self, x: torch.cat([forward(self, c) for c in x.split(tile, dim=1)], dim=1)
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply docs/source/distributing_training.md]: Correct, fixed in 3132df9f.

File: examples/sft_qwen3_8b_1m_context/sft_qwen3_8b_1m_context.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            "rope_parameters": {
+                "rope_type": "yarn",
+                "rope_theta": 1_000_000,
+                "factor": 32.0,
+                "original_max_position_embeddings": 32768,
+            },
+        },
+    )
+
+    trainer = SFTTrainer(model=MODEL, args=training_args, train_dataset=build_dataset())
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/sft_qwen3_8b_1m_context/sft_qwen3_8b_1m_context.py:112]: Not a bug. Checked it under a real FSDP2 launch, 2 ranks:

```
distributed_type=<DistributedType.MULTI_GPU: 'MULTI_GPU'> branch_fires=True
into_create device_map=None keys=['device_map', 'dtype', 'trust_remote_code']
after_create hf_device_map=None devices=['cpu']
```

The premise is that FSDP misses the `["MULTI_GPU", "DEEPSPEED"]` check. It doesn't: at the point the trainer reads it, `args.distributed_state.distributed_type` is still MULTI_GPU, not FSDP. So `device_map` is None, `hf_device_map` is None, and the weights land on CPU with `fsdp_cpu_ram_efficient_loading` intact.

File: examples/sft_qwen3_8b_1m_context/sft_qwen3_8b_1m_context.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+The accelerate config sets `fsdp_activation_checkpointing_offload`, which is what keeps an 8B model
+inside 80 GB at this length. That option is not released yet, so until
+https://github.com/huggingface/accelerate/pull/4175 lands, install accelerate from its branch:
+
+pip install git+https://github.com/huggingface/accelerate.git@fsdp2-activation-offload
+
+accelerate launch \
+    --config_file examples/sft_qwen3_8b_1m_context/context_parallel_8gpu.yaml \
+    examples/sft_qwen3_8b_1m_context/sft_qwen3_8b_1m_context.py
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply examples/sft_qwen3_8b_1m_context/sft_qwen3_8b_1m_context.py:34]: Correct, fixed in the example.

Confirmed against 4.57.1: Qwen3Config has no `rope_parameters` (it's `rope_scaling` + `rope_theta` there), and `PretrainedConfig.from_dict` only applies a kwarg `if hasattr(config, key)`, so the override is dropped with no error.

`accelerate launch` doesn't read the `# /// script` header, so the `transformers>=5.0` it declares isn't enforced at run time. Added an explicit version check in the example instead: a 606 s/step job should not start off silently wrong.

File: docs/source/long_context_training.md
Comment (qgallouedec) [inline docs/source/long_context_training.md:1]: most of this is moved content. I want to re-write the whole section in a follow-up pr

**Review by kashif — APPROVED** [review]

**Review by adithya-s-k — APPROVED** [review]

File: docs/source/long_context_training.md
Comment (AmineDiro) [reply docs/source/long_context_training.md:1]: Do you mean you'll move this file ? I'll still review this

File: docs/source/long_context_training.md [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+| **Backend** | PyTorch FSDP2 | DeepSpeed ZeRO |
+| **Attention** | SDPA only | Flash Attention 2 or SDPA |
+| **Minimum Accelerate** | 1.11.0+ | 1.12.0+ |
+| **Minimum DeepSpeed** | N/A | 0.18.1+ |
+| **Sequence Divisibility** | `cp_size * 2` | `sp_size` |
+| **Zero Stage** | N/A | ZeRO Stage 1/2/3 |
+
+**Ring Attention is better when:**
+- You need to handle extremely long sequences (1M+ tokens)
+- The model has limited attention heads (Ring Attention is not constrained by head count)
```
Comment (AmineDiro) [inline docs/source/long_context_training.md:60]: maybe point to the ulysses sp documentation or paper ?

File: docs/source/long_context_training.md [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+- You want flexibility in scaling to any sequence length
+- Network topology is limited (Ring Attention works with simple P2P ring communication)
+
+**Ulysses is better when:**
+- You have high-bandwidth, low-latency interconnects (NVLink, InfiniBand)
+- The model has many attention heads that can be split across GPUs
+- You want lower communication volume
+- You want faster training speed for moderate sequence lengths (up to ~500k tokens)
```
Comment (AmineDiro) [inline docs/source/long_context_training.md:68]: Ulysses is also better when your model has very special attention like DSA, lInear attention etc...

File: docs/source/long_context_training.md [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+For a detailed comparison, see the [Ulysses and Ring Attention blog post](https://huggingface.co/blog/exploding-gradients/ulysses-ring-attention).
+
+### Ring Attention Implementation (FSDP2)
+
+Ring Attention uses a ring-like communication pattern where each GPU processes a portion of the sequence and passes information to the next GPU in the ring.
+
+#### Requirements and Limitations
+
+1. **Accelerate 1.11.0 or higher** is required for Ring Attention / Context Parallelism support
+2. **FSDP2 (PyTorch FSDP v2)** is required as the distributed training backend
```
Comment (AmineDiro) [inline docs/source/long_context_training.md:84]: any idea why  FSDP2 is required for CP 🤔 ?

Comment (qgallouedec) [reply docs/source/long_context_training.md:84]: because it's only integrated with FSDP 2, and deep speed doesn't plan to integrate it

File: docs/source/long_context_training.md (hunk 9 above)
Comment (qgallouedec) [reply docs/source/long_context_training.md:68]: yes, mentioned in the table in the next pr
